gadget/gadget.go

```diff
@@ -993,3 +994,65 @@ func SystemDefaults(gadgetDefaults map[string]map[string]interface{}) map[string
 	}
 	return nil
 }
+
+// See https://www.kernel.org/doc/html/latest/admin-guide/kernel-parameters.html
+var disallowedKernelArguments = []string{
+	"root", "nfsroot",
+	"init",
+}
+
+// isKernelArgumentAllowed checks whether the kernel command line argument is
+// allowed. Prohibits all arguments listed explicitly in
+// disallowedKernelArguments list and those prefixed with snapd, with exception
+// of snapd.debug. All other arguments are allowed.
+func isKernelArgumentAllowed(arg string) bool {
+	if strings.HasPrefix(arg, "snapd") && arg != "snapd.debug" {
+		return false
+	}
+	if strutil.ListContains(disallowedKernelArguments, arg) {
+		return false
+	}
+	return true
+}
+
+var ErrNoKernelCommandline = errors.New("no kernel command line in the gadget")
+
+// KernelCommandLineFromGadget returns the desired kernel command line provided by the
+// gadget. The full flag indicates whether the gadget provides a full command
+// line or just the extra parameters that will be appended to the static ones.
+// An ErrNoKernelCommandline is returned when thea gadget does not set any
+// kernel command line.
+func KernelCommandLineFromGadget(snapf snap.Container) (cmdline string, full bool, err error) {
+	contentExtra, err := snapf.ReadFile("cmdline.extra")
+	if err != nil && !os.IsNotExist(err) {
+		return "", false, err
+	}
+	contentFull, err := snapf.ReadFile("cmdline.full")
+	if err != nil && !os.IsNotExist(err) {
+		return "", false, err
+	}
+	content := contentExtra
+	whichFile := "cmdline.extra"
+	switch {
+	case contentExtra != nil && contentFull != nil:
+		return "", false, fmt.Errorf("cannot support both extra and full kernel command lines")
+	case contentExtra == nil && contentFull == nil:
+		return "", false, ErrNoKernelCommandline
+	case contentFull != nil:
+		content = contentFull
+		whichFile = "cmdline.full"
+		full = true
+	}
+	cleaned := strings.TrimSpace(string(content))
+	kargs, err := osutil.KernelCommandLineSplit(cleaned)
+	if err != nil {
+		return "", full, fmt.Errorf("invalid kernel command line %q in %v: %v", cleaned, whichFile, err)
```

Comment:
I still tend to agree that returning `full` here isn't super useful, but it also isn't hurting anything so this is fine